Common/BinaryComparisonExtensions.cs

```diff
@@ -0,0 +1,188 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq.Expressions;
+
+namespace QuantConnect
+{
+    /// <summary>
+    /// Provides convenience extension methods for applying a <see cref="BinaryComparison"/> to collections.
+    /// </summary>
+    public static class BinaryComparisonExtensions
+    {
+        /// <summary>
+        /// Filters the provided <paramref name="values"/> according to this <see cref="BinaryComparison"/>
+        /// and the specified <paramref name="reference"/> value. The <paramref name="reference"/> value is
+        /// used as the RIGHT side of the binary comparison. Consider the binary comparison is LessThan and
+        /// we call Filter(values, 42). We're looking for keys that are less than 42.
+        /// </summary>
+        public static TCollection Filter<T, TCollection>(
+            this BinaryComparison comparison,
+            TCollection values,
+            T reference
+            )
+            where TCollection : ICollection<T>, new()
+        {
+            var result = new TCollection();
+            var evaluator = comparison.GetEvaluator<T>();
```

Comment:
The evaluator is a `Func<T, T, bool>` -- it's essentially a delegate that performs the binary comparison. At construction time we could resolve the evaluator and it would save us a dictionary lookup -- is that what you're implying? You say _it could be evaluated at construction time_ -- but at construction time we don't have the arguments, and in fact, for each evaluator we resolve, it will perform many evaluations with many different arguments.

Reply:
We can't do that because we don't know the type `T` -- the good part though is in many of these cases the JIT can heavily optimize. By using the the static `OfType<T>` class we prevent needing a dictionary lookup and then the switch in some cases can be in-lined where the compiler can prove a value. I think this is good for now.

Reply:
hmmm... we could do _something_ to directly tie to the particular slot, but I'm not sure it would end up being any faster and would add some complexity. Would rather prove this it slowing things down via benchmarking before adding to the complexity here.